### DISCUSSION TITLE
Linter - API design

### DISCUSSION THREAD

**Discussion by KFlash** [kataw/kataw#88] [Ideas]
The basic idea for the linter is to add an `autoFix` option to the parser and let the parser fix  "fixable" errors during parsing, and eventually report an diagnostic . Some "errors" are an easy fix such as double quote vs single quote - it only require to swap the bitwise masks before the printer print the CST node.

Implementing a linter this way  eliminates the need for an external walker, and maintain perf. 

There are things to consider before implementing this

1. How should the API look like?  It can't be part of the main parser option because it will be too many options.

2. How could we implement additional rules?

**Comment by KFlash** [comment]
cc / @aladdin-add

**Comment by aladdin-add** [comment]
I would suggest the similar API as eslint (if possible) :)

All of this depends on what's the goal of the linter.
* if it's to be a opinioned linter(like jslint), it can provide very few options.
* if it's to be more customizable and flexibility, a config system is needed. like the eslint rules : https://eslint.org/docs/user-guide/configuring/rules

the public API can be something like
```js
const config = {
  globals: {},
  rules: {},
};
const options = {
  fix: true/false,
  filename: string,
  //...
};

kataw.lint(code, config, options);
```

just my 2 cents! :)

**Reply by KFlash** [reply]
yes, this should be possible with how Kataw is now. Just run it as an linter (it still parse and print the source).
But doing it the ESLint way is very expensive. Still in the future it's possible to extend to custom rules.

**Comment by KFlash** [comment]
@aladdin-add See https://github.com/kataw/kataw/pull/89

**Reply by KFlash** [reply]
@aladdin-add This is the differences

1. Parse the code 
2. Walk the AST / CST
3. Run ESLint
4. Correct errors and walk again
5. Print

Kataw does it like this

1. Parse the code
2. Report errors, lint issues and warnings
3. Print

So you will generate a state like `99 errors,  1 lint issue, 19 warnings` in one go

**Comment by KFlash** [comment]
Kataw are undergoing a mayor refactoring (*not released yet*) so the linter config and options has now changed.

The `files` is all `js` files found, and will be handled by the compiler (*not released yet*)

Each of this `js`  files will have the same set of rules as in the `kataw.json` configuration file and
linting will be done if it's enabled in the config.

The `extends` property works the same as in `ESLint`. I think that is what @aladdin-add wanted?

**This is Kataw's config file:**

```ts
{
    "files": ["**/*.js"],
    "module": false,
    "typecheck": false,
    "linter": {
        "enable": true,
        "extends": [
            "kataw:recommended"
        ],
        "rules": {},
        "rulesDirectory": []
    },
    "minify": {
        "enable": false,
        "options": {}
    },
    "prettify": {
        "enable": true,
        "options": {}
    }
}
```
